ush/biasCorrectionReader.py

```diff
@@ -0,0 +1,317 @@
+    """
+    # Convert string to datetime object
+    dt_object = datetime.strptime(cycle, '%Y%m%d%H')
+
+    # Convert datetime object to Unix epoch time (seconds since January 1, 1970)
+    epoch_time = [int(dt_object.timestamp())]
+
+    return epoch_time
+
+
+def calculate_omf(filename, groups, variable, channels, outgroups, bias_corr, bias_groups, logger):
+    """
+    Calculate observation minus forecast for data that is bias corrected and not bias corrected.
+
+    Parameters:
+        filename (str): input path and filename to netCDF file to be read
+        groups (list): list of strings that contains group names from input .nc file
+        variable (str): variable within input .nc file groups
+        channels (list): channels where data should be grabbed
+        outgroups (list): list of strings that contains the group names for the output .nc file
+        bias_corr (bool): boolean value to determine if bias correction was used or not
+        bias_groups (list): list of observation bias values to be used from input .nc file
+        logger : logging variable that tracks progress of source code
+    Return:
+        return_dict (dict): dictionary of observation minus forecast values
+    """
+    # Add 'ObsValue' to groups to calculate omf
+    groups.append('ObsValue')
+
+    # Grab data from .nc file
+    data_dict = read_ncfile(filename, groups, variable, channels, logger)
+
+    # Calculate omf
+    return_dict = {}
+
+    for i, output in enumerate(outgroups):
+        return_dict[output] = {}
+
+        hofx = groups[i]
+        omf = data_dict['ObsValue'][variable] - data_dict[hofx][variable]
+
+        if bias_corr:
+            return_dict[output]['data'] = omf
+        else:
+            bias_dict = read_ncfile(filename, bias_groups, variable, channels, logger)
+            obsbias = bias_groups[i]
+
+            return_dict[output]['data'] = omf - bias_dict[obsbias][variable]
+
+    return return_dict
+
+
+def calculate_penalty(filename, groups, variable, channels, outgroups, logger):
+    """
+    Calculates penalty by dividing observation minus forecast by effective error.
+
+    Parameters:
+        filename (str): input path and filename to netCDF file to be read
+        groups (list): list of strings that contains group names from input .nc file
+        variable (str): variable within input .nc file groups
+        channels (list): channels where data should be grabbed
+        outgroups (list): list of strings that contains the group names for the output .nc file
+        logger : logging variable that tracks progress of source code
+    Return:
+        return_dict (dict): dictionary of penalty values
+    """
+    return_dict = {}
+
+    omf_dict = calculate_omf(filename, groups, variable, channels, outgroups, bias_corr=True,
+                             bias_groups=None, logger=logger)
+
+    # logic check to see how many effective errors you need?
+    efferr = ['EffectiveError0', 'EffectiveError1']
```

Comment:
Why is this hard coded to 0 and 1 here?

Reply:
I need to think a bit. My main concern is 'what about EffectiveError2?' Recall I mentioned on Thursday if we reproduce the operational GDAS, we would have 0, 1, and 2 due to the two outer loops that run.

Reply:
We can specify it in the yaml again, it just unfortunately adds another conditional in `main()` and then the factory becomes useless.

Reply:
This has to be consistent with the OmF data, right? Why not just construct them consistently with the group numbers? hofx1 -> EffectiveError1 -> EffectiveQC1, etc.

Reply:
That could work. I'll give it a try